frontend/src/metabase/query_builder/components/notebook/steps/JoinStep.jsx

```diff
@@ -603,10 +627,20 @@ class JoinDimensionPicker extends React.Component {
             fieldOptions={options}
             table={query.table()}
             query={query}
-            onFieldChange={field => {
-              onChange(field);
+            onFieldChange={(field, item) => {
+              if (this.hasTemporalUnit(field)) {
+                // When we set a default temporal unit by just clicking the dimension name,
+                // FieldList provides a second argument to onFieldChange callback
+                // When a temporal unit is selected explicitly (like by Week, Quarter, Year, etc.)
+                // the item argument is undefined
+                // In this case we want to make sure the second dimension will use the same unit
+                onChange(field, { overwrite: !item });
```

Comment:
does `item` here mean "sub-dimension" or does it mean more than that? I see that the word `item` is used in `DimensionOptions.js` and `DimensionList.jsx` too so I'm just sort of confused.

I'm left wondering if there isn't a more generic feature hidden here. Specifically, the passing of subdimensions to `Join` which then reconciles whatever is passed to it. IDK if that makes sense--just food for thought.

Reply:
After a fresh look here it seems to be too hacky TBH. I changed `DimensionList` so it passes a `{ isSubDimension: true }` object as a second argument if subdimension is selected (it used to pass only the first arg). This should be more straightforward